src/core/lib/promise/detail/basic_join.h

```diff
@@ -0,0 +1,189 @@
+// Copyright 2021 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GRPC_CORE_LIB_PROMISE_DETAIL_BASIC_JOIN_H
+#define GRPC_CORE_LIB_PROMISE_DETAIL_BASIC_JOIN_H
+
+#include <grpc/impl/codegen/port_platform.h>
+
+#include <utility>
+#include "absl/utility/utility.h"
+#include "src/core/lib/gprpp/bitset.h"
+#include "src/core/lib/gprpp/construct_destruct.h"
+#include "src/core/lib/promise/detail/promise_factory.h"
+#include "src/core/lib/promise/poll.h"
+
+namespace grpc_core {
+namespace promise_detail {
+
+// This union can either be a functor, or the result of the functor (after
+// mapping via a trait). Allows us to remember the result of one joined functor
+// until the rest are ready.
+template <typename Traits, typename F>
+union Fused {
+  explicit Fused(F&& f) : f(std::forward<F>(f)) {}
+  explicit Fused(PromiseLike<F>&& f) : f(std::forward<PromiseLike<F>>(f)) {}
+  ~Fused() {}
+  // Wrap the functor in a PromiseLike to handle immediately returning functors
+  // and the like.
+  using Promise = PromiseLike<F>;
+  GPR_NO_UNIQUE_ADDRESS Promise f;
+  // Compute the result type: We take the result of the promise, and pass it via
+  // our traits, so that, for example, TryJoin and take a StatusOr<T> and just
+  // store a T.
+  using Result = typename Traits::template ResultType<typename Promise::Result>;
+  GPR_NO_UNIQUE_ADDRESS Result result;
+};
+
+// A join gets composed of joints... these are just wrappers around a Fused for
+// their data, with some machinery as methods to get the system working.
+template <typename Traits, size_t kRemaining, typename... Fs>
+struct Joint : public Joint<Traits, kRemaining - 1, Fs...> {
+  // The index into Fs for this Joint
+  static constexpr size_t kIdx = sizeof...(Fs) - kRemaining;
+  // The next join (the one we derive from)
+  using NextJoint = Joint<Traits, kRemaining - 1, Fs...>;
+  // From Fs, extract the functor for this joint.
+  using F = typename std::tuple_element<kIdx, std::tuple<Fs...>>::type;
```

Comment:
I think you're also missing #include <tuple> here ;-)

Reply:
Done